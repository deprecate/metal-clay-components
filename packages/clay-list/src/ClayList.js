import 'clay-dropdown';
import 'clay-icon';
import 'clay-label';
import 'clay-link';
import 'clay-sticker';
import './ClayListItem';

import Component from 'metal-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';
import {Config} from 'metal-state';

import templates from './ClayList.soy.js';

/**
 * Metal ClayList component.
 */
class ClayList extends Component {
	/**
	 * Continues the propagation of the checkbox changed event
	 * @param {!Event} event
	 * @private
	 */
	handleItemToggled_(event) {
		this.emit('itemToggled', event);
	}
}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayList.STATE = {
	/**
	 * CSS classes to be applied to the element.
	 * @instance
	 * @memberof ClayList
	 * @type {?string|undefined}
	 * @default undefined
	 */
	elementClasses: Config.string(),

	/**
	 * Id to be applied to the element.
	 * @instance
	 * @memberof ClayList
	 * @type {?string|undefined}
	 * @default undefined
	 */
	id: Config.string(),

	/**
	 * List of items.
	 * @instance
	 * @memberof ClayList
	 * @type {?array|undefined}
	 * @default undefined
	 */
	items: Config.array(),

	/**
	 * Flag to indicate if the list group items are selectable.
	 * @instance
	 * @memberof ClayList
	 * @type {?bool|undefined}
	 * @default false
	 */
	selectable: Config.bool().value(false),

	/**
	 * Schema mapping list item fields with item data properties.
	 * @instance
	 * @memberof ClayList
	 * @type {!object}
	 * @default undefined
	 */
	schema: Config.shapeOf({
		actionItems: Config.string(),
		description: Config.string(),
		href: Config.string(),
		icon: Config.string(),
		iconShapesMap: Config.object(),
		iconsMap: Config.object(),
		labels: Config.string(),
		labelsStylesMap: Config.object(),
		selected: Config.string(),
		title: Config.string(),
	}).required(),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @instance
	 * @memberof ClayList
	 * @type {?string|undefined}
	 * @default undefined
	 */
	spritemap: Config.string(),
};

defineWebComponent('clay-list', ClayList);

Soy.register(ClayList, templates);

export {ClayList};
export default ClayList;
