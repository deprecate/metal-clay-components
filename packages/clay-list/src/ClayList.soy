{namespace ClayList}

/**
 * This renders the component's whole content.
 */
{template .render}
	{@param schema: list<[
		contentRenderer: string,
		fieldName: string,
		fieldsMap: ?,
		iconsMap: ?,
		iconShapesMap: ?,
		labelStylesMap: ?
	]>}
	{@param? elementClasses: string}
	{@param? handleItemToggled_: any}
	{@param? id: string}
	{@param? items: list<?>}
	{@param? selectable: bool}
	{@param? spritemap: string}

	{let $attributes kind="attributes"}
		class="list-group show-quick-actions-on-hover
			{if $elementClasses}
				{sp}elementClasses
			{/if}
		"

		{if $id}
			id="{$id}"
		{/if}
	{/let}

	<ul {$attributes}>
		{if $items}
			{call .items}
				{param handleItemToggled_: $handleItemToggled_ /}
				{param items: $items /}
				{param schema:	$schema /}
				{param selectable: $selectable /}
				{param spritemap: $spritemap /}
			{/call}
		{/if}
	</ul>
{/template}

{template .item}
	{@param item: ?}
	{@param schema: list<[
		contentRenderer: string,
		fieldName: string,
		fieldsMap: ?,
		iconsMap: ?,
		iconShapesMap: ?,
		labelStylesMap: ?
	]>}
	{@param? handleItemToggled_: any}
	{@param? selectable: bool}
	{@param? spritemap: string}

	{let $listItemClasses kind="text"}
    list-group-item list-group-item-flex
  		{if $item.selected}
	  		{sp}active
		  {/if}
	{/let}

	<li class="{$listItemClasses}">
		{if $selectable}
			<div class="flex-col">
				{call ClayCheckbox.render}
					{param checked: $item.selected /}
					{param events: ['change': $handleItemToggled_] /}
					{param hideLabel: true /}
					{param label: 'select' /}
				{/call}
			</div>
		{/if}

		{foreach $fieldSchema in $schema}
			{delcall ClayList.Field variant="$fieldSchema.contentRenderer"}
				{param fieldSchema: $fieldSchema /}
				{param item: $item /}
				{param spritemap: $spritemap /}
				{param value: $item[$fieldSchema.fieldName] /}
			{/delcall}
		{/foreach}

		{if $item.actionItems and $spritemap}
			<div class="flex-col">
				{call .quickMenu}
					{param items: $item.actionItems /}
					{param spritemap: $spritemap /}
				{/call}

				{call ClayActionsDropdown.render}
					{param items: $item.actionItems /}
					{param spritemap: $spritemap /}
				{/call}
			</div>
		{/if}
	</li>
{/template}

/**
 * This renders all the items in the list, iteraing over the items and its
 * possible nested items.
 */
{template .items}
	{@param items: list<?>}
	{@param schema: list<[
		contentRenderer: string,
		fieldName: string,
		fieldsMap: ?,
		iconsMap: ?,
		iconShapesMap: ?,
		labelStylesMap: ?
	]>}
	{@param? handleItemToggled_: any}
	{@param? selectable: bool}
	{@param? spritemap: string}

	{foreach $item in $items}
		{if $item.items}
			<li class="list-group-header">
				<h3 class="list-group-header-title">{$item.label}</h3>
			</li>

			{call .items}
				{param handleItemToggled_: $handleItemToggled_ /}
				{param items: $item.items /}
				{param schema:	$schema /}
				{param selectable: $selectable /}
				{param spritemap: $spritemap /}
			{/call}
		{else}
			{call .item}
				{param handleItemToggled_: $handleItemToggled_ /}
				{param item: $item /}
				{param schema: $schema /}
				{param selectable: $selectable /}
				{param spritemap: $spritemap /}
			{/call}
		{/if}
	{/foreach}
{/template}

/**
 * This renders the quick actions menu
 */
{template .quickMenu}
	{@param items: list<[
		disabled: bool,
		href: string,
		icon: string,
		label: string,
		quickAction: bool,
		separator: bool
	]>}
	{@param spritemap: string}

	<div class="quick-action-menu">
		{foreach $item in $items}
			{if $item.quickAction and $item.icon and $spritemap}
				{call ClayLink.render}
					{param elementClasses: 'quick-action-item' /}
					{param href: $item.href /}
					{param icon: $item.icon /}
					{param spritemap: $spritemap /}
				{/call}
			{/if}
		{/foreach}
	</div>
{/template}

/**
 * This renders the content of the item.
 */
{deltemplate ClayList.Field variant="'content'"}
	{@param fieldSchema: [
		contentRenderer: string,
		fieldName: string,
		fieldsMap: ?,
		iconsMap: ?,
		iconShapesMap: ?,
		labelStylesMap: ?
	]}
	{@param item: ?}
	{@param value: ?}
	{@param? spritemap: string}

	<div class="flex-col flex-col-expand">
		<h4 class="list-group-title text-truncate">
			{if $fieldSchema.fieldsMap and $fieldSchema.fieldsMap.href and $item[$fieldSchema.fieldsMap.href]}
				{call ClayLink.render}
					{param href: $item[$fieldSchema.fieldsMap.href] /}
					{param label: $item[$fieldSchema.fieldsMap.title] /}
				{/call}
			{else}
				{$item[$fieldSchema.fieldsMap.title]}
			{/if}
		</h4>

		{if $fieldSchema.fieldsMap and $fieldSchema.fieldsMap.description and $item[$fieldSchema.fieldsMap.description]}
			<p class="list-group-subtitle text-truncate">{$item[$fieldSchema.fieldsMap.description]}</p>
		{/if}

		{if $fieldSchema.fieldsMap and $fieldSchema.fieldsMap.label and $item[$fieldSchema.fieldsMap.label]}
			<div class="list-group-detail">
				{call ClayLabel.render}
					{param label: $item[$fieldSchema.fieldsMap.label] /}
					{param style: $fieldSchema.labelStylesMap ? $fieldSchema.labelStylesMap[$item[$fieldSchema.fieldsMap.label]] ?: $fieldSchema.labelStylesMap['*'] : null /}
				{/call}
			</div>
		{/if}
	</div>
{/deltemplate}

/**
 * This renders the icon of the item.
 */
{deltemplate ClayList.Field variant="'icon'"}
	{@param fieldSchema: [
		contentRenderer: string,
		fieldName: string,
		fieldsMap: ?,
		iconsMap: ?,
		iconShapesMap: ?,
		labelStylesMap: ?
	]}
	{@param item: ?}
	{@param value: ?}
	{@param? spritemap: string}

	<div class="flex-col">
		{if $spritemap}
			{call ClaySticker.render}
				{param icon: $fieldSchema.iconsMap ? $fieldSchema.iconsMap[$value] ?: $fieldSchema.iconsMap['*'] : $value /}
				{param shape: $fieldSchema.iconShapesMap ? $fieldSchema.iconShapesMap[$value] ?: $fieldSchema.iconShapesMap['*'] : null /}
				{param spritemap: $spritemap /}
			{/call}
		{/if}
	</div>
{/deltemplate}

/**
 * This renders a simple content of the item.
 */
{deltemplate ClayList.Field variant="'simple'"}
	{@param fieldSchema: [
		contentRenderer: string,
		fieldName: string,
		fieldsMap: ?,
		iconsMap: ?,
		iconShapesMap: ?,
		labelStylesMap: ?
	]}
	{@param item: ?}
	{@param value: ?}
	{@param? spritemap: string}

	{if $fieldSchema.fieldsMap and $fieldSchema.fieldsMap.href and $item[$fieldSchema.fieldsMap.href]}
		{call ClayLink.render}
			{param href: $item[$fieldSchema.fieldsMap.href] /}
			{param label: $item[$fieldSchema.fieldsMap.title] /}
		{/call}
	{else}
		{$item[$fieldSchema.fieldsMap.title]}
	{/if}
{/deltemplate}