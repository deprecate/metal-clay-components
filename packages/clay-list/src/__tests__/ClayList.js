import ClayList from '../ClayList';

let component;

describe('ClayList', function() {
	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('should render the default markup', () => {
		component = new ClayList({
			schema: {},
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayList with classes', () => {
		component = new ClayList({
			elementClasses: 'my-custom-class',
			schema: [],
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayList with id', () => {
		component = new ClayList({
			id: 'myId',
			schema: [],
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a selectable ClayList', () => {
		component = new ClayList({
			items: [
				{
					title: 'Star Wars',
				},
			],
			schema: {
				title: 'title',
			},
			selectable: true,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a selectable ClayList with groups', () => {
		component = new ClayList({
			items: [
				{
					items: [
						{
							title: 'Star Wars',
						},
					],
					label: 'Movies',
				},
			],
			schema: {
				title: 'title',
			},
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayList and emit an event on item toggled', () => {
		const spy = jest.fn();

		component = new ClayList({
			events: {itemToggled: spy},
			items: [
				{
					title: 'Star Wars',
				},
			],
			schema: {
				title: 'title',
			},
			selectable: true,
		});

		component.element.querySelector('input[type="checkbox"]').click();

		expect(spy).toHaveBeenCalled();
	});

	it('should fail when no schema is passed', function() {
		expect(() => {
			component = new ClayList({
				items: [],
			});
		}).toThrow();
	});
});
