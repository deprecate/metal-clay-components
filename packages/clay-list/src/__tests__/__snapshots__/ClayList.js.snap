// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayList should render a ClayList with classes 1`] = `<ul class="list-group show-quick-actions-on-hover elementClasses my-custom-class"></ul>`;

exports[`ClayList should render a ClayList with id 1`] = `<ul class="list-group show-quick-actions-on-hover" id="myId"></ul>`;

exports[`ClayList should render a selectable ClayList 1`] = `
<ul class="list-group show-quick-actions-on-hover">
  <li class="list-group-item list-group-item-flex">
    <div class="flex-col">
      <div class="custom-control custom-checkbox">
        <label>
          <input class="custom-control-input" ref="input" type="checkbox">
          <span class="custom-control-indicator"></span>
          <span class="custom-control-description sr-only">select</span>
        </label>
      </div>
    </div>
    <div class="flex-col flex-col-expand">
      <h4 class="list-group-title text-truncate">Star Wars</h4>
    </div>
  </li>
</ul>
`;

exports[`ClayList should render a selectable ClayList with groups 1`] = `
<ul class="list-group show-quick-actions-on-hover">
  <li class="list-group-header">
    <h3 class="list-group-header-title">Movies</h3>
  </li>
  <li class="list-group-item list-group-item-flex">Star Wars</li>
</ul>
`;

exports[`ClayList should render the default markup 1`] = `<ul class="list-group show-quick-actions-on-hover"></ul>`;
